README.md
<!-- badges: start -->
[![CRAN
status](https://www.r-pkg.org/badges/version/SLEMI)](https://cran.r-project.org/package=SLEMI)
[![R-CMD-check](https://github.com/TJetka/SLEMI/workflows/R-CMD-check/badge.svg)](https://github.com/TJetka/SLEMI/actions)
[![Codecov test coverage](https://codecov.io/gh/TJetka/SLEMI/branch/master/graph/badge.svg)](https://codecov.io/gh/TJetka/SLEMI?branch=master)
<!-- badges: end -->

# Statistical Learning based Estimation of Mutual Information (SLEMI)
 The R package SLEMI is designed to estimate channel capacity between finite state input and multidimensional output from experimental data. For efficient computations, it uses iterative algorithm based on logistic regression.  The core function `capacity_logreg_main()` is the basic interface to all functionalities provided in the package. A comprehensive documentation is available in directory [`old_vignettes/SLEMI_vignette.pdf`](https://github.com/TJetka/SLEMI/blob/master/old_vignettes/SLEMI_vignette.pdf).
 ## Setup
 ### Requirements - Hardware
  + A 32 or 64 bit processor (recommended: 64bit)
  + 1GHz processor (recommended: multicore for a comprehensive analysis)
  + 2GB MB RAM (recommended: 4GB+, depends on the size of experimental data)
 ### Requirements - Software
The main software requirement is the installation of the R environment (version: >= 3.6), which can be downloaded from [R project website](https://www.r-project.org) and is distributed for all common operating systems.  We tested the package in R environment installed on Windows 7, 10; Mac OS X 10.11 - 10.13 and Ubuntu 18.04 with no significant differences in the performance. The use of a dedicated Integrated development environment (IDE), e.g. [RStudio](https://www.rstudio.com) is recommended. 
 Apart from base installation of R, SLEMI requires following packages:
 1. for installation 
   + devtools
  
2. for estimation
  
  + e1071
  + Hmisc
  + nnet
  + glmnet
  + caret
  + doParallel (if parallel computation are needed)
  
3. for visualisation
  + ggplot2
  + ggthemes
  + gridExtra
  + corrplot
  
4. for data handling
  + reshape2
  + stringr
  + plyr
 Each of the above packages can be installed by running 
> `install.packages("name_of_a_package")`
 in the R console.
 ### Installation
SLEMI is on R CRAN now. Please install using
>`install.packages("SLEMI")`

 In order to install directly from GitHub, use following commands in R's console
 > `# install.packages("devtools") # run if not installed`

 > `library(devtools)`
 > `install_github("TJetka/SLEMI")`
 
 ## Basic usage
 The package is based on a main wrapper function - `capacity_logreg_main()` for calculation of channel capacity, which calls specific methods implemented within this package. Similarly, functions `mi_logreg_main()` can be used to estimate mutual information, while `prob_discr_pairwise()` to compute probabilities of discrimination between two different input states.
 ### Preparing data
 For the calculation of channel capacity between X and Y you need structure experimental data into a single `data.frame` object with observations in rows, one column with values of input (X), preferably of `factor` type and columns with measured output (Y) of `numeric` type.
 ### Run
 In order to estimate channel capacity, using basic logistic regression model, call
> `capacity_logreg_main(dataRaw, signal, response, output_path)`
where: 
* `dataRaw` is a data.frame with experimental data as described above
* `signal` is a character indicating the name of column in `dataRaw` with the input (X)
* `response` is a character vector indicating names of columns in `dataRaw` with output (Y) variables
* `output_path` is a character with the directory, to which results of the estimation should be saved
 ### Results
 The function `capacity_logreg_main` returns a list, whose main elements are
 * cc - channel capacity estimate (in bits)
* p_opt - numeric vector with the optimal input distribution
* model - `nnet` object describing fitted logistic regression model
 For convenience of further analysis, this list is saved in `output_path` directory in a file `output.rds`. In addition to that, a set of exploratory graphs are created to visualise obtained estimates.
 ## Examples
 Additional examples of using package with some background on information theory is given in [`paper/TestingProcedures.pdf`](https://github.com/TJetka/SLEMI/blob/master/paper/TestingProcedures.pdf) and implemented in script [`paper/testing_procedures.R`](https://github.com/TJetka/SLEMI/blob/master/paper/testing_procedures.R). Codes used in publication are accessible from [`paper/paper_MP.R`](https://github.com/TJetka/SLEMI/blob/master/paper/paper_MP.R) and [`paper/paper_SI.R`](https://github.com/TJetka/SLEMI/blob/master/paper/paper_SI.R) respectively.
 ### Datasets
 In the manuscript describing methodological aspects of our algorithm we present the analysis of information transmission in NfkB pathway upon the stimulation of TNF-$\alpha$. Experimental data from this experiment in the form of single-cell time series are attached to the package as a data.frame object and can be accessed using `data_nfkb` variable.
 Each row of `data_nfkb` represents a single observation of a cell. Column 'signal' indicates the level of TNF-$\alpha$ stimulation for a given cell, while columns 'response_T', gives the normalised ratio of nuclear and cytoplasmic transcription factor as described in Supplementary Methods of the corresponding publication. 
 ## Other functionalities
 ### Additional parameters
 Apart from required arguments, the function `capacity_logreg_main` has also other parameters than can be used to tune the activity of the algorithm. These are
 * `model_out` (`default=TRUE`) - logical, specify if `nnet` model object should be saved into output file
* `graphs` (`default=TRUE`) - logical, controls creating diagnostic plots in the output directory.
* `plot_width` (`default = 6`) - numeric, the basic width of created plots 
* `plot_height` (`default = 4`) - numeric, the basic height of created plots
* `scale` (`default = TRUE`) - logical, value indicating if the columns of `dataRaw` are to be centered and scaled, what is usually recommended for the purpose of stability of numerical computations. From a purely theoretical perspective, such transformation does not influence the value of channel capacity.
* `lr_maxit` (`default = 1000`) - (argument of `nnet` package) a maximum number of iterations of optimisation step in logistic regression algorithm. Set to higher value if your data is more complex or of high dimension.
* `MaxNWts` (`default = 5000`) - (argument of `nnet` package) a maximum number of parameters in logistic regression model. Set to higher value if you data has many dimensions or input has many states.
 ### Diagnostic procedures
 We implemented two diagnostic procedures to control the performance of channel capacity estimation and to measure  uncertainty due to finite sample size and model over-fitting. These include:
 1. Bootstrap test - capacity is re-calculated using $x$% of data, sampled from original dataset without replacement. After repeating procedure $n$ times, its standard deviation can be treated as an error of original estimate.
2. Over-fitting test - original data is divided into Training and Testing datasets. Then, logistic regression is estimated using $x$% of data (training dataset) and integrals of channel capacity are calculated via Monte Carlo using remaining $(1-x)$% of data (testing dataset). It is repeated $n$ times.
 In order to use those procedures, user must provide additional arguments to function `logreg_capacity_main()`, i.e.
 * testing (default=FALSE) - a logical value that turn on/off testing mode,
* TestingSeed (default= 1234) - the seed for the random number generator used to sample original dataset,
* testing_cores (default= 4) - a number of cores to use (via `doParallel` package) in parallel computing, 
* boot_num (default= 40) - a number of repetitions of the bootstrap , 
* boot_prob (default= 0.8) - a fraction of initial observations to  use in the bootstrap, 
* traintest_num (default= 40) - a number of repetitions of the overfitting test,
* partition_trainfrac (default= 0.6) - a fraction of initial observations to use as a training dataset in the overfitting test
 ## Support
 Please mail t.jetka at gmail.com in case of any bugs, problems and questions regarding package or inquiries regarding information theory.
 ## Reference
 Please cite
> Jetka T, Nienałtowski K, Winarski T, Błoński S, Komorowski M (2019) Information-theoretic analysis of multivariate single-cell signaling responses. PLOS Computational Biology 15(7): e1007132. https://doi.org/10.1371/journal.pcbi.1007132
 ## License
 SLEMI is released under the GNU license and is freely available. A comprehensive documentation is available in directory [`old_vignettes/SLEMI_vignette.pdf`](https://github.com/TJetka/SLEMI/blob/master/old_vignettes/SLEMI_vignette.pdf).


NEWS.md
# SLEMI 1.0.1

* Added a `NEWS.md` file to track changes to the package.
* Moved the home repository to github.com/TJetka (from github.com/sysbiosig)
* Fixed the bug related to the new version of dependency: ggplot2 (stat_summary function)
* Added unit tests for package maintenance

---

# SLEMI 1.0.0

* Initial CRAN submission

R/mi_logreg_main.R
#' Estimate mutual information between discrete input and continuous output
#'
#' The main wrapping function for basic usage of SLEMI package for estimation of mutual information. Firstly, data is pre-processed
#' (all arguments are checked, observation with NAs are removed, variables are scaled and centered (if scale=TRUE)). Then basic estimation is carried out
#' and (if testing=TRUE) diagnostic tests are computed. If output directory path is given (output_path is not NULL), graphs visualising the data and the analysis
#' are saved there, together with a compressed output object (as .rds file) with full estimation results.
#'
#' In a typical experiment aimed to quantify information flow a given signaling system, input values \eqn{x_1\leq x_2 \ldots... \leq x_m}, ranging from 0 to saturation are considered.
#' Then, for each input level, \eqn{x_i}, \eqn{n_i} observations are collected, which are represented as vectors 
#' \deqn{y^i_j \sim P(Y|X = x_i)}
#' Within information theory the degree of information transmission is measured as the mutual information
#' \deqn{MI(X,Y) = \sum_{i=1}^{m} P(x_i)\int_{R^k} P(y|X = x_i)log_2\frac{P(y|X = x_i)}{P(y)}dy,}
#' where \eqn{P(y)} is the marginal distribution of the output. MI is expressed in bits and \eqn{2^{MI}} can be interpreted as the number of 
#' inputs that the system can resolve on average.
#'
#' In contrast to existing approaches, instead of estimating, possibly highly dimensional, conditional output distributions \eqn{P(Y|X =x_i)}, we propose to estimate the discrete, conditional input distribution, 
#' \eqn{P(x_i |Y = y)}, which is known to be a simpler problem. Estimation of the MI using estimates of \eqn{P(x_i |Y = y)}, denoted here as \eqn{\hat{P}(x_i|Y = y)}, is possible as the MI, can be
#' alternatively written as
#' \deqn{MI(X,Y) = \sum_{i=1}^{m} P(x_i)\int_{R^k} P(y|X = x_i)log_2\frac{P(x_i|Y = y)}{P(x_i)}dy}
#' The expected value (as in above expression) with respect to distribution \eqn{P(Y|X = x_i)} can be approximated by the average with respect to data
#' \deqn{MI(X,Y) \approx \sum_{i=1}^{m} P(x_i)\frac{1}{n_i} \sum_{j=1}^{n_i} P(y|X = x_i)log_2\frac{\hat{P}(x_i|Y = y^i_j)}{P(x_i)}dy}
#' Here, we propose to use logistic regression as \eqn{\hat{P}(x_i|Y = y^i_j)}. Specifically,
#' \deqn{log\frac{P(x_i |Y = y)}{P(x_m|Y = y)} \approx \alpha_i +\beta_iy}
#'
#' Additional parameters: lr_maxit and maxNWts are the same as in definition of multinom function from nnet package. An alternative
#' model formula (using formula_string arguments) should be provided if  data are not suitable for description by logistic regression
#' (recommended only for advanced users). Preliminary scaling of  data (argument scale) should be used similarly as in other 
#' data-driven approaches, e.g. if response variables are comparable, scaling (scale=FALSE) can be omitted, while if they represent 
#' different phenomenon (varying by units and/or magnitude) scaling is recommended.
#'
#' @section References:
#' [1] Jetka T, Nienaltowski K, Winarski T, Blonski S, Komorowski M,  
#' Information-theoretic analysis of multivariate single-cell signaling responses using SLEMI,
#' \emph{PLoS Comput Biol}, 15(7): e1007132, 2019, https://doi.org/10.1371/journal.pcbi.1007132.
#'
#'
#' @param dataRaw must be a data.frame object
#' @param signal is a character object with names of columns of dataRaw to be treated as channel's input.
#' @param response is a character vector with names of columns of dataRaw  to be treated as channel's output
#' @param output_path is the directory in which output will be saved
#' @param side_variables (optional) is a character vector that indicates side variables' columns of data, if NULL no side variables are included
#' @param pinput is a numeric vector with prior probabilities of the input values. Uniform distribution is assumed as default (pinput=NULL).
#' @param formula_string (optional) is a character object that includes a formula syntax to use in logistic regression model. 
#' If NULL, a standard additive model of response variables is assumed. Only for advanced users.
#' @param lr_maxit is a maximum number of iteration of fitting algorithm of logistic regression. Default is 1000.
#' @param MaxNWts is a maximum acceptable number of weights in logistic regression algorithm. Default is 5000.
#' @param plot_height -  basic dimensions (height) of plots, in inches
#' @param plot_width -  basic dimensions (width) of plots, in inches
#' @param model_out is the logical indicating if the calculated logistic regression model should be included in output list
#' @param data_out  is the logical indicating if the data should be included in output list
#' @param scale is a logical indicating if the response variables should be scaled and centered before fitting logistic regression
#' @param testing is the logical indicating if the testing procedures should be executed
#' @param TestingSeed is the seed for random number generator used in testing procedures
#' @param boot_num is the number of bootstrap tests to be performed. Default is 10, but it is recommended to use at least 50 for reliable estimates.
#' @param boot_prob is the proportion of initial size of data to be used in bootstrap
#' @param testing_cores - number of cores to be used in parallel computing (via doParallel package)
#' @param sidevar_num is the number of re-shuffling tests of side variables to be performed. Default is 10, but it is recommended to use at least 50 for reliable estimates.
#' @param traintest_num is the number of overfitting tests to be performed. Default is 10, but it is recommended to use at least 50 for reliable estimates.
#' @param partition_trainfrac is the fraction of data to be used as a training dataset
#'
#' @return a list with several elements:
#' \itemize{
#' \item output$regression - confusion matrix of logistic regression predictions
#' \item output$mi         - mutual information in bits
#' \item output$model      - nnet object describing logistic regression model (if model_out=TRUE)
#' \item output$params     - parameters used in algorithm
#' \item output$time       - computation time of calculations
#' \item output$testing    - a 2- or 4-element output list of testing procedures (if testing=TRUE)
#' \item output$testing_pv - one-sided p-values of testing procedures (if testing=TRUE)
#' \item output$data       - raw data used in analysis
#' }
#' @export
#' @examples 
#' tempdata=data_example1
#' outputCLR1=mi_logreg_main(dataRaw=tempdata, signal="signal", response="response")
#' 
#' tempdata=data_example2
#' outputCLR2=mi_logreg_main(dataRaw=tempdata, signal="signal", response=c("X1","X2","X3")) 
#' 
#' #For further details see vignette
mi_logreg_main<-function(dataRaw, signal="input", response=NULL,output_path=NULL,
  side_variables=NULL,pinput=NULL, formula_string=NULL,
                                          lr_maxit=1000, MaxNWts = 5000, 
                                          testing=FALSE, model_out=TRUE,scale=TRUE,
                                          TestingSeed=1234,testing_cores=1,
                                          boot_num=10,boot_prob=0.8,
                                          sidevar_num=10,
                                          traintest_num=10,partition_trainfrac=0.6,
                                          plot_width=6,plot_height=4,
                                          data_out=FALSE){
  
 if(!is.null(output_path)){
    dir.create(output_path,recursive=TRUE)
 }

  #Debugging:
  message("Estimating mutual information ...")
  
  time_start=proc.time()
  dataRaw=as.data.frame(dataRaw)
  
  #
  if (is.null(response)){
    response=paste0("output_",1:(ncol(dataRaw)-1) )  
  }
  
  # checking assumptions
  if (is.null(output_path)) { 
    message('Output Path is not defined. Graphs and RDS file will not be saved.')
    }
  if (!is.data.frame(dataRaw)) {
    stop('data is not in data.frame format')
  }
  if ( sum(colnames(dataRaw)==signal)==0 ) {
    stop('There is no column described as signal in data')
  }
  if (!sum(colnames(dataRaw) %in% response)==length(response) ) {
    stop('There is no column described as response in data')
  }
  if (!is.null(side_variables)){
    if ( !sum(colnames(dataRaw) %in% side_variables)==length(side_variables) ) {
    stop('There is no column described as side_variables in data')
    }
  }

  
   data0=dataRaw[,c(signal,response,side_variables)]
   if ( any(apply(data0,1,function(x) any(is.na(x)) )) ) {
     message(" There are NA in observations - removing ")
     data0=data0[!apply(data0,1,function(x) any(is.na(x)) ),]
   }
  
   data0=func_signal_transform(data0,signal)
   tempcolnames=colnames(data0)
   tempsignal=data.frame(data0[,(tempcolnames%in%c(signal,paste(signal,"_RAW",sep="") ) )])
   colnames(tempsignal)<-tempcolnames[(tempcolnames%in%c(signal,paste(signal,"_RAW",sep="") ) )]
   data0=data.frame(data0[,!(tempcolnames%in%c(signal,paste(signal,"_RAW",sep="") ) )])
   colnames(data0)<-tempcolnames[!(tempcolnames%in%c(signal,paste(signal,"_RAW",sep="") ) )]
   

  
  #PreProcessing
  temp_idnumeric=sapply(data0,is.numeric)
  if (scale&sum(temp_idnumeric)==1) {
    data0[,temp_idnumeric]<-(data0[,temp_idnumeric]-mean(data0[,temp_idnumeric]))/stats::sd(data0[,temp_idnumeric])
    data <- cbind(data0,tempsignal)
  } else if (scale) {
    preProcValues <- caret::preProcess(data0, method = c("center", "scale"))
    data <- cbind(stats::predict(preProcValues, data0),tempsignal)
  } else {
    data <- cbind(data0,tempsignal)
  }
  rm(temp_idnumeric)

  #Debugging:
  #cat("... completed")
  
  output<-mi_logreg_algorithm(data=data,signal=signal,response=response,side_variables=side_variables,
                              pinput=pinput,formula_string=formula_string, model_out = model_out,
                                              lr_maxit=lr_maxit,MaxNWts =MaxNWts) 


  
  if (testing){
      output$testing<-mi_logreg_testing(data,signal=signal,response=response,side_variables=side_variables,
                                                      lr_maxit=lr_maxit,MaxNWts =MaxNWts,pinput=pinput,
                                                      formula_string=formula_string,
                                                      TestingSeed=TestingSeed,testing_cores=testing_cores,
                                                      boot_num=boot_num,boot_prob=boot_prob,
                                                      sidevar_num=sidevar_num,
                                                      traintest_num=traintest_num,partition_trainfrac=partition_trainfrac)

      
      output$testing_pv<-lapply(output$testing,function(x){
        tmp_boot_cc=sapply(x,function(xx) xx$mi)
        c(mean(tmp_boot_cc<output$mi),mean(tmp_boot_cc>output$mi))
      })
  }
  
  output$time   <- proc.time() - time_start
  output$params <- c(lr_maxit=lr_maxit,MaxNWts =MaxNWts)
  
  if (data_out){
    output$data   <- dataRaw
  }
  
  if(!is.null(output_path)){
    dir.create(output_path,recursive=TRUE)

    message("Drawing graphs and saving objects...")
    temp_logGraphs=try(output_graphs_main(data=dataRaw,signal=signal,response=response,side_variables=side_variables,cc_output=output,
                                output_path=output_path,height=plot_height,width=plot_width),silent=FALSE)
    rm(temp_logGraphs)
    #Debugging:
    #cat("... finished")

    saveRDS(output,file=paste(output_path,'output.rds',sep=""))
    message(paste0("Estimation finished. Results saved in ",output_path,""))

    } else {
      message(paste0("Estimation finished."))
    }
  
  output
}
